var/spack/repos/builtin/packages/py-ipympl/package.py

```diff
@@ -15,16 +15,32 @@ class PyIpympl(PythonPackage):
 
     license("BSD-3-Clause")
 
+    version("0.9.4", sha256="cfb53c5b4fcbcee6d18f095eecfc6c6c474303d5b744e72cc66e7a2804708907")
     version("0.8.8", sha256="5bf5d780b07fafe7924922ac6b2f3abd22721f341e5e196b3b82737dfbd0e1c9")
 
-    depends_on("py-setuptools@40.8:", type="build")
-    depends_on("py-ipython@:8", type=("build", "run"))
-    depends_on("py-numpy", type=("build", "run"))
-    depends_on("py-ipython-genutils", type=("build", "run"))
-    depends_on("pil", type=("build", "run"))
-    depends_on("py-traitlets@:5", type=("build", "run"))
-    depends_on("py-ipywidgets@7.6:7", type=("build", "run"))
-    depends_on("py-matplotlib@2:3", type=("build", "run"))
-    depends_on("py-jupyter-packaging@0.7", type="build")
-    depends_on("py-jupyterlab@3", type="build")
-    depends_on("yarn", type="build")
+    with default_args(type="build"):
+        with when("@0.9:"):
+            depends_on("py-hatchling")
```

Comment:
The build failure report that I submitted shows that `pip` fails because `opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm install` does not find Node.js:
```py
  ValueError: Please install Node.js and npm before continuing installation. You may be able to install Node.js from your package manager, from conda, or directly from the Node.js website (https://nodejs.org).
  Traceback (most recent call last):
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm', 'install']' returned non-zero exit status 1.
```
The concrete spec from the build failure report shows that `nodejs` is not in the `spack` build environment of this build.

To check, I did `git switch develop;git pull upstream develop`, re-ran the `spack install` and this then built `nodejs` as part of some `"build"` dependency in `develop`. The `spack install` on `develop` failed for a different reason (but that is off-topic, I think):
```py
  >> 105    node_modules/@@rjsf/utils/lib/getDateElementProps.d.ts(1,15): error TS1005: ',' expected.
  >> 106      node_modules/@@rjsf/utils/lib/index.d.ts(17,36): error TS1005: ',' expected.
  >> 107      node_modules/@@types/lodash/common/common.d.ts(262,65): error TS1005: '?' expected.
  >> 108      node_modules/@@types/lodash/common/object.d.ts(1026,46): error TS1005: '?' expected.
  >> 109      node_modules/@@types/lodash/common/object.d.ts(1031,46): error TS1005: '?' expected.
  >> 110      node_modules/@@types/lodash/common/object.d.ts(1041,46): error TS1005: '?' expected.
```
So, then, switched back to this PR's branch and re-ran the failed `spack install`,
but no change, `nodejs`, despite ready for re-use, was still not included in the build.

I assume that `nodejs` should be a `"run"` dependency of `opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm`, so I checked if this fixes the build:
```suggestion
            depends_on("py-hatchling")
            depends_on("node-js")
```
This does fix the build failure. It apparently triggered on that host, as it has no `nodejs` installed in the VM's OS.

However, I guess that the correct fix location for re-adding a `"run"` dependency on `nodejs` is likely either `py-jupyterlab` or another dependency of this build.

I'm sure you know better than  me how to fix this correctly.

Reply:
That's odd, because I don't have nodejs/npm/yarn installed but was able to install ipympl without issue on macOS. Can you show me the full error message stack trace?

Reply:
@adamjstewart <details><summary>Please click on this line to unfold/fold: output from build of py-iplmpl with stack strace

</summary>

```py
[+] /usr (external glibc-2.35-spkgpfgkrsiezsjbgpao5m6ddj2hchhj)
[+] /usr (external tar-1.34-dsrra4ynbpxpkhph65dgedqxzykk4axk)
[+] /usr (external curl-7.81.0-2fanoa6ijem63ldfn4s4c52enblwxe5m)
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/gcc-runtime-13.1.0-25qo3ld7klulv5juopmjnk73qte743xz
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libiconv-1.17-jxe63kwxqplstnvcueiggoke5v2ezi6d
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/openblas-0.3.28-xrdrjso7wej7ivc4g2jkplsl3o5273jh
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/qhull-2020.2-svxch26m67cyq7rq2emqou7nmeiev2qq
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libffi-3.4.6-r6emmhbggetnzptutbkk7qffbsmezppz
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/zlib-ng-2.2.1-somd2q767lkqx6je2pz6am2gztdbovfe
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libsodium-1.0.20-msqegfakuhtoi5m6w6vjle7jfdip7zjp
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libxcrypt-4.4.35-dhqsmbqdm53fnmk6ynzaef6msaunah6r
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libyaml-0.2.5-v5gpzid2d7je2cs3c3ypafqpk6viyauk
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/bzip2-1.0.8-geajalrno4mt3uers3ji7ytbkk6qvkez
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libmd-1.0.4-z4llf26xjzstptwwlm4efpzljny4hb47
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/pcre-8.45-lcdzy6wxv7d7t3ho65cg7233npatwtim
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/xz-5.4.6-pwjjfgxqlhfn26tycgwy2iezdpzmt7br
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/util-linux-uuid-2.40.2-tyzqiklai55oisaoewn5ro5nv3wy7mcv
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/ncurses-6.5-5iauhqykhluyoehuhrwhvvuyvpud2765
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libjpeg-turbo-3.0.3-jem7g7qag3qr3fjecpjgca7kt3ahsx63
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libpng-1.6.39-b2xcvuqrq3nba5dssxhp3r5ovmre524x
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/openssl-3.3.1-qbjwgb6gct3qnc5hpufmmffieve7dkib
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libbsd-0.12.2-calffvojasv62v5v6b4mplsycwbj67fs
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libxml2-2.10.3-2cbjv2cwl4mm5bxmrwhiif6ghxrsf75w
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/readline-8.2-w6hn7avdihv3xjsswtwougplip4ulmgu
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/freetype-2.13.2-mghqys465mualtb72huzyebs4tyx3mhk
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libssh2-1.11.0-3cjisru5qmkeayahagvbd3kx5aanjc22
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libzmq-4.3.5-rzwwh3wvf55gt7pq4lx3qsmpnr4kfdrw
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/expat-2.6.3-u2bucaey36mxkx6zbov2tk46okm3zsh2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/gettext-0.22.5-ib3eyquni4uqvocpgo7wwaejwtdty2hl
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/sqlite-3.46.0-i6xusas3hfs57j5p72at6gucksqd3gsw
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/gdbm-1.24-434jch2p64j3el672tchjaemupijit4w
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/libgit2-1.8.0-lwbwgvjb2rfn6x2hapjaks7hsg62j5t3
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-3.11.9-djketg7p3skmfq64434e36uqxbs5qo2u
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/rust-1.81.0-djdnu6r7meyo4jybsem5bsqoubjns3y7
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-json5-0.9.14-3li26u6ktaihq22ge5usrzoe7nfnur7u
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-soupsieve-2.4.1-p4vt3xhg7tafsz5aaqr5hg6npbceksmp
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pathspec-0.11.1-nujvlhnuax2os5b5cb6bbxhhlduhwpak
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-certifi-2023.7.22-dbvcb6h2v5kwa5fhaum5dzcvg3ur55x6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jsonpointer-2.0-jkusvy4h3s7l6yfpkbpk4vyif4qdwdun
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pluggy-1.5.0-zh5zhmixbmnhhh3cdngphq3ge7xs22qf
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ptyprocess-0.7.0-zbpi2jbym3mt4ug2viqhw7ynoyyg5ozq
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-fqdn-1.5.1-bcnr2lc4qijd5hhacyqvctfeka4r2p35
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-wheel-0.41.2-zk7lxqzh652c3wce523gx2elbxy4jpk6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-setuptools-63.4.3-nhwlcgms42f6wy2jhhkwehw52aiaj5vm
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pexpect-4.9.0-kj2nky3bjc3uoovqk4yhnfky3abyhry2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-idna-3.4-llwuqf246c6grlchmhc3jmm6cwl3vexm
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-urllib3-2.1.0-llpckciukbc5dvntyds6ogaboyqkq6vi
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-uri-template-1.2.0-zdhstupzbbjtgig57e56xtx752kutzdl
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pyyaml-6.0.2-x2h7j7w7oh3zqbo6woye4ocpx22lyovs
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-executing-1.2.0-afbowq2zmchz6pwj3535ghf6vgrns3hm
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-sniffio-1.3.0-hiadfuajmqbearfkaelzcmnkk4ilalce
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pyparsing-3.1.2-ru6uia7qg4yqpkieigwqmv7jyzul6tzx
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-markupsafe-2.1.3-n4cu7zd5pxowjzrhjwb3umrzwojk446m
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-wcwidth-0.2.7-44tkuluwajgxcwetc65czdgw5b26wx5a
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-attrs-23.1.0-3onvpfqgfmgpmdmybbbil6ov67b3btih
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-numpy-1.25.2-3elf6xqqyttbxu7yym3elv6hcde2qsd5
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-overrides-7.3.1-i7ners4evo52xex672wflo6u2mn4rgwz
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pandocfilters-1.5.0-r2s7vyw5b7ekbe4mzg5mf5qawamcxl2i
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-traitlets-5.14.3-ytwrtom3i6ibeehs2vlgj4yvt3hrxlv2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-psutil-5.9.5-mqg3m22ctlyvlxs7q6wn3bokiul5gfpg
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-widgetsnbextension-4.0.3-v5b7w3ofvntum26c5mngrw6jcco3sqlp
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-beautifulsoup4-4.12.3-srino6tgoibnbzcjwdxvyjsldbho5tiw
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-packaging-24.1-h42e7ltjw7f4j7d7envusrkm2ojuchwf
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pure-eval-0.2.2-isfrvsg5vqhvjb7rmt4bqeyfzkuq3ert
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-send2trash-1.8.3-f4zlv4bzxcbsg4g4kew54nouv47jwkt5
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-kiwisolver-1.4.5-syz35n6kxuqamhqgymsdcg3j7icsr3cz
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-python-json-logger-2.0.7-hxbkehdcqzwqmjncjuftoinqo3hbwfmb
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-defusedxml-0.7.1-2uaxgpakt2gq6maqc5gi6lk35qjuorxr
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pybind11-2.13.5-dlkq7co7zt7zcnfhop7d72yulu3qr2hf
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-six-1.16.0-urodnl2q76gul3azynck7m26lefkiyx2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pycparser-2.21-rtughkhilmlvq3uvtxd6xifrqzrqm7ww
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-platformdirs-3.10.0-gi63jbgdsmzr57toc4y6hhl3fbjleceg
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-debugpy-1.6.7-rhhig2b4sq3pxzk77gugvtzj5lw6gpfw
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-webcolors-24.6.0-pwjyg37tedr3btfll7nwv64fpsoh457b
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-nest-asyncio-1.6.0-ezyexl4wr5ozdrvb4cdjb6rtjngypnci
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-fastjsonschema-2.16.3-odqic4pchehl4ujb3yb2xcz2ki3hdnto
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipython-genutils-0.2.0-5i3kxkvpwprzvsjime73dqbgma5fv7qo
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-webencodings-0.5.1-sbfqzzsestosg5ajkkbse6oeo6nxlxpi
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-rfc3986-validator-0.1.1-c43mnflcbqrvafgmonkr5wiriznyzdd6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-zope-interface-7.0.3-6je5krffebyqpecpx6ube2q2zi2gwnhr
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-rpds-py-0.20.0-ylbnns42zveq26lczccyrr4qhesngutn
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-trove-classifiers-2023.8.7-3qv2wnbd2ehhqntw2unqarucl246mkcj
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-decorator-5.1.1-b7fduu56f32ufwvczmja4ocioeb2k42e
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-fonttools-4.39.4-3r7ue7xvcrd64jqs3blkag5iixr3coym
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-greenlet-2.0.2-wibfv4k4jvfoekm3zv4kwetsfdfyozgt
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-tornado-6.3.3-7issrh5pdbqqhr3vumcjbln7rplxz4xw
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-types-python-dateutil-2.8.19.14-wog2aznkdpt4qtmp4idqpr3atilk6jy6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-typing-extensions-4.12.2-5qjccwoyg4blvdiantw5blqwcm5elfdc
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-widgets-3.0.3-2utyvv2giuislbokek3bmg7rxarr3547
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-websocket-client-1.8.0-nxwpixqerwik554idmurxzk3guwzcmwe
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-babel-2.15.0-y2m6o54y4mrwtvw44psrhndq6w4755h3
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-editables-0.5-s5il65owynbhdgeug3yoexxqm7xiuuvq
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pillow-11.0.0-ezztobwqwdnsjuvv743qyedm7vxadidf
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-cycler-0.11.0-qskcgghg3jdqdynyxg2apw6izsedpw4z
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pygments-2.18.0-dbmw74xybnrlfitfwxj5mi52feprp77r
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-prometheus-client-0.17.0-yct5fm3rzbcljqd6e4dadksr7faibsjq
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-charset-normalizer-3.3.0-5wbqgvtri5fjrlszfbtqbr3yyhncb6is
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-parso-0.8.3-gfc724pdj3nuudt5ysvgrcirpk3gya7e
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-zope-event-5.0-wk64txrhrka5cafyhlnfgv4kzhvey23z
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-async-lru-1.0.3-mr2464oo7qxgtkhlui4buwdf3sxoge6t
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-mistune-2.0.5-ifvswysiaeywjcle4dotyjpog6cdxhbg
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-anyio-4.0.0-mxhshbr46uyajbv6prnv3swc7plynjfi
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jinja2-3.1.4-uuxxbfyr57bgd4tvey54hhstduppxnpv
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-prompt-toolkit-3.0.43-qgfl4h3fhc7f2qaqiy27muk4tiqmbb63
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-comm-0.1.4-x7pfrxkkr3ieenwwherljj7xpl6nbsgb
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-contourpy-1.3.0-m6c2bny73ljt2pcwws4acj5zx3nz3ae2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-python-dateutil-2.8.2-oiyhzav2ndbwgjt5bdsywnoztezqfl5r
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-rfc3339-validator-0.1.4-dlhyxwsbqvy2ks5wedokoyxuir7h3re2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-asttokens-2.4.0-nq3aacsh2fpxwcwempwqvltwvran7qf5
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-cffi-1.17.1-jgiakq5l4h5uvnbejgmetv7m4ewath32
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-core-5.3.0-klhd6dusn5zgj7wqglodzy2fwuicmshp
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-tinycss2-1.2.1-kdaerd3q4azmfsvec37pxrdkub6xesqj
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-bleach-6.0.0-icekh7wvlcduik2kdx6szfneikjytr3e
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-referencing-0.35.1-raxxvmyzz7dh3nhjrk7e4z35rhevngrg
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-terminado-0.17.1-xs4q5qfdsnccnirjdpsc2gvr6ex5owho
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatchling-1.25.0-neajfoqmju4mpfltohu7pxkkvtzj2qyl
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-pygments-0.2.2-omw5ak6lfekaaqws2fsw4jdgk6nztvox
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-requests-2.32.3-g6gdizlkvfskvi7fhukxkak2hcfqlmxk
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jedi-0.18.2-5dhdnbkrzdxuyba3ywb6iuvpebkd2ib5
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-arrow-1.3.0-2i5uvmuhgm76ajncaixcgzfjzjx3rbhv
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-matplotlib-3.9.2-mxpht3yio2ydk53y4gvsjo335pdhaedx
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-stack-data-0.6.2-3p5gdzo2ojzi4zo3qe5u37g3peg4sxku
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-gevent-23.7.0-gghu6ihuwx3yxurj3jinwtoq2mumjpvx
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-argon2-cffi-bindings-21.2.0-mch6vbzzr7aqbybq2425mzloemznq6zm
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-jupyter-builder-0.8.3-7ilth3oeqvaeujd7yzmsafgevxfhowu2
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jsonschema-specifications-2023.12.1-3oyfztxi2ilbkclf6jbnlrzyjw2uixag
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-server-terminals-0.4.4-e5vnpli4sa7dgxzrrhcgtahrk3425vqn
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-nodejs-version-0.3.2-huihhwggksxw5632phqelakvpa2itqqy
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-isoduration-20.11.0-tn4ftzd7qvn7ijpsakil6y4la3kfa3du
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-matplotlib-inline-0.1.6-dldh77rrabvf7i5dx2evuawwucn5zw6m
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pyzmq-25.0.2-zindwdxpdyjbrdlvua2776nahvgmh26l
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-argon2-cffi-21.3.0-4e2k63szofhlcq3x4ilpwmjaj3hacm7z
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jsonschema-4.22.0-bdijq2v7fkxlsoqba24rppcry53jvmta
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipython-8.28.0-mxzqva2t7cdblqvwmgctgmaksvvmkd27
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-client-8.2.0-x6dmdtrcjrykxd7ifzxofkfidgb6fn3k
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-nbformat-5.8.0-idxxkg3ofp5rpd6wtptkrcdvhdf7hpi6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-events-0.10.0-g75x3dljsdba4w33h5xutb4gvphlvx5o
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipykernel-6.29.5-kpvlwc5ip2qjsgmepi4jhgoeyiqkgbuv
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-nbclient-0.8.0-dagd726wjyjioxcchrmzwla7o5qnpuzk
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipywidgets-8.0.2-z7o2csfonobewhsl4rnj36zfwsk4oahn
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-nbconvert-7.14.1-hhg3a3kbaajq22qyqlw3hn3luyfbedst
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-server-2.14.2-c55js2i2hzng7e3uzmk4mjwtfuo7dcb3
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-notebook-shim-0.2.3-fceajtbggskkbgl3qrrag552ofceiaks
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyter-lsp-2.2.0-wqqxgzcajlgayaikhyxzzhvydyxd33i6
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-server-2.22.1-xivrtoigh7ty75ra7bogxk2zzq4vepd3
[+] opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo
==> Installing py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg [149/149]
==> No binary for py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg found: installing from source
==> Fetching https://files.pythonhosted.org/packages/source/i/ipympl/ipympl-0.9.4.tar.gz
==> No patches needed for py-ipympl
==> py-ipympl: Executing phase: 'install'
==> [2024-10-24-22:47:05.817252] 'opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k/bin/python3' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg' '.'
Using pip 23.1.2 from opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip (python 3.11)
Non-user install due to --prefix or --target option
Ignoring indexes: https://pypi.org/simple
Created temporary directory: /tmp/pip-build-tracker-u5deegp1
Initialized build tracking at /tmp/pip-build-tracker-u5deegp1
Created build tracker: /tmp/pip-build-tracker-u5deegp1
Entered build tracker: /tmp/pip-build-tracker-u5deegp1
Created temporary directory: /tmp/pip-install-ta_j3ai1
Created temporary directory: /tmp/pip-ephem-wheel-cache-b5b0063f
Processing .stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-src
  Added file://.stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-src to build tracker '/tmp/pip-build-tracker-u5deegp1'
  Created temporary directory: /tmp/pip-modern-metadata-mhlwy1yv
  Preparing metadata (pyproject.toml): started
  Running command Preparing metadata (pyproject.toml)
  INFO:hatch_jupyter_builder.utils:Running jupyter-builder
  INFO:hatch_jupyter_builder.utils:Building with hatch_jupyter_builder.npm_builder
  INFO:hatch_jupyter_builder.utils:With kwargs: {'path': '.', 'build_cmd': 'build:prod', 'npm': ['jlpm']}
  INFO:hatch_jupyter_builder.utils:Installing build dependencies with npm.  This may take a while...
  INFO:hatch_jupyter_builder.utils:> opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm install
  Traceback (most recent call last):
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm", line 10, in <module>
      sys.exit(main())
               ^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/lib/python3.11/site-packages/jupyterlab/jlpmapp.py", line 43, in main
      execvp("node", ["node", YARN_PATH, *argv])
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/lib/python3.11/site-packages/jupyterlab/jlpmapp.py", line 24, in execvp
      cmd = which(cmd)
            ^^^^^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-server-2.22.1-xivrtoigh7ty75ra7bogxk2zzq4vepd3/lib/python3.11/site-packages/jupyterlab_server/process.py", line 58, in which
      raise ValueError(msg)
  ValueError: Please install Node.js and npm before continuing installation. You may be able to install Node.js from your package manager, from conda, or directly from the Node.js website (https://nodejs.org).
  Traceback (most recent call last):
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
      main()
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 152, in prepare_metadata_for_build_wheel
      whl_basename = backend.build_wheel(metadata_directory, config_settings)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatchling-1.25.0-neajfoqmju4mpfltohu7pxkkvtzj2qyl/lib/python3.11/site-packages/hatchling/build.py", line 58, in build_wheel
      return os.path.basename(next(builder.build(directory=wheel_directory, versions=['standard'])))
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatchling-1.25.0-neajfoqmju4mpfltohu7pxkkvtzj2qyl/lib/python3.11/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
      build_hook.initialize(version, build_data)
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-jupyter-builder-0.8.3-7ilth3oeqvaeujd7yzmsafgevxfhowu2/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 87, in initialize
      raise e
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-jupyter-builder-0.8.3-7ilth3oeqvaeujd7yzmsafgevxfhowu2/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 82, in initialize
      build_func(self.target_name, version, **build_kwargs)
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-jupyter-builder-0.8.3-7ilth3oeqvaeujd7yzmsafgevxfhowu2/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 113, in npm_builder
      run([*npm_cmd, "install"], cwd=str(abs_path))
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-hatch-jupyter-builder-0.8.3-7ilth3oeqvaeujd7yzmsafgevxfhowu2/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 229, in run
      return subprocess.check_call(cmd, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-3.11.9-djketg7p3skmfq64434e36uqxbs5qo2u/lib/python3.11/subprocess.py", line 413, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-jupyterlab-4.0.1-kgt5kf6x3x4h7wb2cw3fhinljzkbf3qo/bin/jlpm', 'install']' returned non-zero exit status 1.
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k/bin/python3 opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmpib7omts4
  cwd: .stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-src
  Preparing metadata (pyproject.toml): finished with status 'error'
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
Exception information:
Traceback (most recent call last):
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/operations/build/metadata.py", line 35, in generate_metadata
    distinfo_dir = backend.prepare_metadata_for_build_wheel(metadata_dir)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/utils/misc.py", line 713, in prepare_metadata_for_build_wheel
    return super().prepare_metadata_for_build_wheel(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_impl.py", line 186, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_impl.py", line 311, in _call_hook
    self._subprocess_runner(
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/utils/subprocess.py", line 252, in runner
    call_subprocess(
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/utils/subprocess.py", line 224, in call_subprocess
    raise error
pip._internal.exceptions.InstallationSubprocessError: Preparing metadata (pyproject.toml) exited with 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 169, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 248, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
                      ^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 73, in resolve
    collected = self.factory.collect_root_requirements(root_reqs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 491, in collect_root_requirements
    req = self._make_requirement_from_install_req(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 453, in _make_requirement_from_install_req
    cand = self._make_candidate_from_link(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 206, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
                                       ^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 293, in __init__
    super().__init__(
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 156, in __init__
    self.dist = self._prepare()
                ^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 304, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 516, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 631, in _prepare_linked_requirement
    dist = _get_prepared_distribution(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 69, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/distributions/sdist.py", line 61, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/req/req_install.py", line 555, in prepare_metadata
    self.metadata_directory = generate_metadata(
                              ^^^^^^^^^^^^^^^^^^
  File "opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/site-packages/pip/_internal/operations/build/metadata.py", line 37, in generate_metadata
    raise MetadataGenerationFailed(package_details=details) from error
pip._internal.exceptions.MetadataGenerationFailed: metadata generation failed
Removed file://.stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-src from build tracker '/tmp/pip-build-tracker-u5deegp1'
Removed build tracker: '/tmp/pip-build-tracker-u5deegp1'
==> Error: ProcessError: Command exited with status 1:
    'opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k/bin/python3' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg' '.'

1 error found in build log:
     64      ╰─> See above for output.
     65
     66      note: This error originates from a subprocess, and is likely not a problem with pip.
     67      full command: opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k/bin/p
           ython3 opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-pip-23.1.2-rwutgrhpohc24fyl24lkmrpgshsoxo3p/lib/python3.11/s
           ite-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmpib7omts4
     68      cwd: .stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-src
     69      Preparing metadata (pyproject.toml): finished with status 'error'
  >> 70    error: metadata-generation-failed
     71
     72    × Encountered error while generating package metadata.
     73    ╰─> See above for output.
     74
     75    note: This is an issue with the package mentioned above, not pip.
     76    hint: See above for details.

See build log for details:
  .stage/spack-stage-py-ipympl-0.9.4-hsh7kiighiucschvpyku4r3zbr4fwblg/spack-build-out.txt

==> Error: Terminating after first install failure: ProcessError: Command exited with status 1:
    'opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/python-venv-1.0-caywfzkj2wb5f5kyrr3kuozihntgpx6k/bin/python3' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/py-ipympl-0.9.4-hsh7kiighiucschvpykubkaindbkaibkbkbkaibkbkaindl@n1:~/spackbkaibkbkaibkbkb
```

</details>

Reply:
Try again with the latest commit? I think you just forgot the npm dependency (it's packaged separately from node-js).